### ISSUE TITLE
what is the aim of next gen of weight? 

### ISSUE THREAD

**Issue by l1t1** [gcp/leela-zero#1625]
is the aim to produce much stronger with weight bigger size or not so much stronger and keep the size increases small ?
and also care about the speed of promotion?
some possible size
192x20
224x20
256x15
256x20

**Comment by gcp** [comment]
We'll do 256x20 next at v=1600.

We can do small increases as parallel training can somewhat point out when a lot is being left on the table by the network being too small.

**Comment by Marcin1960** [comment]
Would it be possible to check 224x15 or 192x20 first, please?

With 256x20 I would have to buy a new computer :(

**Comment by herazul** [comment]
Why would you ? If the network is better at equal think time, the same hardware is gonna give better result.

**Comment by featurecat** [comment]
your computer can't run that size at all? or it is only slow?

**Comment by Marcin1960** [comment]
@herazul & @ferazul

Too slow, too few playouts, drops a few stones down. I checked several times.

**Comment by jkiliani** [comment]
> We'll do 256x20 next at v=1600.

@gcp Sounds great, but what is the current plan regarding this? Are we continuing to test @bjiyxo's networks until one is "enough" stronger to justify switching? Or are you training up a bootstrap net as well?

**Comment by featurecat** [comment]
It will probably be stronger anyway Marcin, with same thinking time.

**Comment by herazul** [comment]
@Marcin1960 What hardware do you have ? I find it weird that your hardware can't run a 20x256 net, even my smartphone can run it with 300+ playouts, that's enough to have a VERY good net strengh (according to dontbtme, even 40 playout is enough to beat leela11, 300 is very strong.) And it's only a smartphone !

**Comment by Marcin1960** [comment]
At 20 seconds I get about 2000 playouts, with 256x20 I get about 900. Probably, it is not enough to use 256x20 potential.

**Comment by Friday9i** [comment]
I just noticed @gcp commented on the issue while I was writing!
Cool, let's go to 256x20 and we'll see if it improves reasonably fast (if I'm wrong in my comment below ;-) or if we should switch quickly to a larger size (if I'm right ; -).
And good news for v=1600, quite convinced it will be efficient! And faster for sure : -)

What I was writing before the last posts: 
Hopefully more than 256x20! Maybe 256x30?
I'm hoping we'll go to a significantly larger network, otherwise it may be quite ineficient.
Here's my understanding of the situation, and why I think we should go around 256x30:
1) The current process allowed us to push LZ's 192x15 network "quite far", ie quite close from it's theoretical maximum!
Indeed, we spend quite a lot of time training it, and **we had a significant help from ELF** which sped up the process **a lot**: we gained around 1000 Elo since ELF appeared and without ELF, it would have take a very long time to reach the current level, possibly 1 year or more if we look at the curve before ELF appeared! For sure, we've got some remaining space to go a bit further, but it's becoming inefficient as it takes a lot of time to improve a little bit.
2) ELF is a significantly larger network (so it has more potential) and it was reasonably well trained, but it is probably further away from it's potential maximum (as the self-improvement was slow). 
There's no reason to believe Facebook rushed to deliver it, as they could have waited a few more weeks to train it if they were expecting some strong improvements: hence, it is "quite well trained". However, it is probably not as close to it's potential as LZ's 192x15 net is, because they didn't get help from a stronger and larger net (LZ had ELF to speed up it's training, ELF hadn't this kind of help). And the progress becoming exponentially slower (unless it gets help from a better net), they stopped it further away from the max, even if they had a lot of processing power. So ELF has some remaining potential, but it would take ages to try to catch the last bits of improvements, even with a lot of processing power as Facebook had.
From 1) and 2), going to 256x20 (very close to the 224x20 from ELF), we would be in the position Facebook was with ELF: progress is very slow and quite frustrating, as it is not too far from it's maximum potential (without external help from a stronger net).
Hence, we should go to a larger network which has more potential, even if it is slower to run! Slower to run, yes, but **at least it has a lot of potential!** Hence, 256x30 seems to be a nice target.. Or why not directly to 256x40?

**Comment by gcp** [comment]
>Sounds great, but what is the current plan regarding this? Are we continuing to test @bjiyxo's networks until one is "enough" stronger to justify switching? Or are you training up a bootstrap net as well?

I'm hoping that @bjiyxo's ones work, though if necessary I can have a try as well.

**Comment by herazul** [comment]
@Marcin1960 900 playout is good. If the new net is better at 900 playout than the old one at 2000, it's worth it, even on same hardware.

@Friday9i I think it's reasonable to think that the 20x256 wont achieve much better that the 20x224 ELF, their net seems to have been pushed close to the limits. I would also prefer to yolo the new 40x256 @bjiyxo's net if it's that good, it would maybe not be better on time parity at first, but we would have a lot more potential of progression !

**Comment by Marcin1960** [comment]
@herazul "900 playout is good. If the new net is better at 900 playout than the old one at 2000, it's worth it, even on same hardware."

It was not my experience so far. But if the new net is really, really better? We would see.

**Comment by herazul** [comment]
@Marcin1960 It's already the case, try matching our last 15x192 net versus the last 10x128 net on time parity : the 15x192 will destroy the 10x128 net, even a 1/10 playouts.
Even a 40*256 net will probably be better on your hardware than our 15x192 on time parity (it will only compute 300 playouts instead of 2000, but it will be better anyway even with that few playouts.

**Comment by Marcin1960** [comment]
@herazul herazul Probably with 40*256 my PC will croak at zero playouts.

What PC do you have?

**Comment by herazul** [comment]
@Marcin1960 A powerful one (intel 7700K - nvidia 1080), but if my smartphone run a 20x256, you computer will run a 40x256, i cant see why not. (or it's a very very weak computer ^^)
What's your hardware ?

**Comment by Marcin1960** [comment]
LeelaZeroT runs on Intel Core Quad 6600 4GB RAM, 32bit Win 7, AMD Radeon HD 7750c 1GB

**Comment by Splee99** [comment]
@marcin1960, have you tried elf weight? How strong is it on your computer?

**Comment by Marcin1960** [comment]
@Splee99  "have you tried elf weight? How strong is it on your computer?"

I did, got weaker. Intel Core Quad 6600 4GB RAM, 32bit Win 7, AMD Radeon HD 7750c 1GB

**Comment by gcp** [comment]
I mean I have nothing against a 256x40 either.

But as we are very close to the 192x15 limit, I need a concrete step on how to continue. For now, going to 256x20 (which we're likely to have a good candidate of) and fixing v to 1600 (see past discussion wrt t=1 making v= closer to p=, and my notes about gating *probably* requiring keeping a reasonable amount of search going). This is a conservative approach that I expect no problems implementing with not too much effort involved, rather consistent with how things were done so far.

If it's racing to get the strongest network, our current setup is probably much too conservative to be good, and mostly serves as a baseline encouraging other groups to beat it (looks like there might be a second ELF?).

**Comment by herazul** [comment]
Yeah it's understandable too.

It depend of the strenght of 40x256, we will have to see how it perform. It could be a solid choice for the next step, we will see soon !

**Comment by BITINT** [comment]
Facebook hinted at the possibility of releaseing a new network soon, maybe after the games. 
if so, it is very likely it will at least be 40b x 256f if not larger.

Thus moving to 40b would be a good idea since it will be a one time last final move for the LZ project most likely (I don't ineviasion LZ getting to 80 b ever etc) 

if moving to a 20b, this means LZ might one day catch up to the first ELF net but will never catch up to the second or third ELF nets.  

Take into account that new gen of graphics cards are coming out soon, then 40b is the reasonable size for the majority of the general public that would be inclined to things like lczero and zeroleela anyway, we can't cater to the most vocal people running subpar machines as the lowest common denominator otherwise this "console" approach will mean nothing ever takes off the ground. There will always be people who can' afford to get a new gfx card to run a larger network size, but the decision should be based and made on and for the greatest good and with the forward thinking approaching that as time goes on, hardware improves, it doesn't stand still...

**Comment by john45678** [comment]
What about Golaxy as a comparison to a 20b LZ, is their a good estimate how strong Golaxy is?

**Comment by herazul** [comment]
We don't have access to golaxy, thus no way to know how our networks performs against it.

**Comment by john45678** [comment]
@herazul  true. Thanks. (and of course hardware etc).

**Comment by Marcin1960** [comment]
@BITINT "people running subpar machines as the lowest common denominator otherwise this "console" approach will mean nothing ever takes off the ground."

What do you mean "nothing will ever takes off the ground."? That the project is a failure?

**Comment by BITINT** [comment]
The project has already succeeded in the context of it being the first free and open source and publicly available (in terms of training data, which ELF OpenGo is NOT) that raises the baseline lower bound to above pro levels (which of course prior to AG was not though possible anytime soon)... 

But since it LZ has already expanded beyond its original initial intended scope (recall it was just to test the bootstrapping and the method, not to grow to 15 block and looking to move to 20b or larger etc) then in the new extended scope, the goal is to see just how far and how larger it can become. Going to just 20b from 15b would be artifically limiting... for this new stage and segment of the project's lifespan.

**Comment by Friday9i** [comment]
The good news is that apparently @gcp has "nothing against 256x40" : -)))
I have the feeling that the practical question slowing down the move to a larger net is "How to get a good starting network, on-par with the current best 192x15 net with time-parity?".
Current 20B nets are not far from this state, but not exactly at it. And we know that 20B nets won't go much higher than the current 192x15 net, as demonstrated by ELF. 
Current 30B or 40B are probably even worse at time parity, as they are much slower and intrinsically not much better **now**. However, their potential is probably much higher: hence, their level should improve fast through selfplay, and after a few weeks and several promotions of new nets, they should go through the roof and become the new indisputed best nets ; -).
Hence, the approach "we wait for an equivalent net at time parity" may be a mistake...: there's little reason not to switch to a larger net, even if it is slightly worse at time-parity in the beginning.

**Comment by PhilipFRipper** [comment]
@gcp I'm usually all for the conservative path here, not that my opinion matters! But I think you and the other coders, and the community with their donated resources, have done a good job setting a conservative baseline, why not try more experimental / fun stuff now?  I think you've done what you promised.  Go ahead now and do whatever interests you, we got your back.

**Comment by Marcin1960** [comment]
"The community with their donated resources, have done a good job setting a conservative baseline, why not try more experimental / fun stuff now?"

This project, especially software development, was fascinating, including kyu level. I learned a lot. 

The new direction - bigger nets and competing for the wins, for me is boring. Better to watch soccer games. I might buy bigger screen. ;)

**Comment by wonderingabout** [comment]
@Marcin1960
"The new direction - bigger nets and competing for the wins, for me is boring. Better to watch soccer games. "
lol !

lz as a project was indeed very interesting so far, and i think no one here doubts this is about time to upscale to larger network size, however what to do from now on and how to do it (20/30/40 blocks ?), is the real question
perhaps this question may become more clear after we see next 20 blocks results
i'm still surprised about how latest lz official network gained 7% of winrate against elf at 400 games in the last 400k games, this still doesnt change the time to upscale network size but i expected it to be more like arround 80% winrate for elf.

is it possible to queue 20b v17 against latest lz official network and compare the winrate with last v17 match ?

**Comment by BITINT** [comment]
one way it could be done is to implement a referendum of sorts, for a week or so, have networks of all sizes, 20b, 30b, 40b, and have a quasi-training sessions to allow anyone to do any size they wish, for the sole purpose of collecting realistic "voting" power, so that the net arch size that actually ends up with the most effective attention/interest will get nominated as the next step up... this puts democracy into practice and gives the volunteers in this crowd funded community the ultimate voice.

**Comment by wonderingabout** [comment]
@BITINT 
my personal opinion on democracy : people need to be ruled, the majority is easy to manipulate, always changes its mind back and forth, and is not wise

**Comment by Ishinoshita** [comment]
>  How to get a good starting network, on-par with the current best 192x15 net with time-parity?

Is time-parity an absolute criterion to chose the network? The strentgh of self-play can be controlled by the number of visits, that could be kept at 3200v if needed. Even is game generation rate was, say, to be halved, since a bigger network has increased learning capability, it may make faster progress even with fewer games per time unit of time. LZ self-play Elo recent progress were very slow, less than 400Elo in 1M games. Maybe 500k games will suffice for a bigger net to cover the same step in strength ;-)

**Comment by Marcin1960** [comment]
OK, decided to give a try to 224x20 net (smaller than 256x20) - 62b5417b with 75% win

I downloaded it, and unfortunately I found that it is 192x15 net. Something got misplaced?

**Comment by Marcin1960** [comment]
224x20 net - 62b5417b with 75% win, looks good. I am doing tuning now and net seems faster than I expected, much less than 2x slowdown. It is 200 ELO above what can correspond to 2x more playouts. So far so good :)

**Comment by betterworld** [comment]
> If it's racing to get the strongest network, our current setup is probably much too conservative to be good

> bigger nets and competing for the wins, for me is boring

So what do people think about doing stuff that the other groups haven't tried yet? Currently everyone is re-implementing Alphago Zero. The power of the "Zero" approach has been shown by now and there are still other things that haven't been researched as thoroughly:


Modifications to the neural network layout:
- Inputs for handicap and komi
- Learn some time management (which moves to play fast and which to play slow)
- Ladder input
- Reward functions for winning by as many points as possible or with as few moves as possible

Other things:
- Add human games to the training database at this stage and continue reinforcement learning

**Comment by featurecat** [comment]
I'd like ladder input. That way we can have a bot of every playing strength that won't lose 50% of its games due to ladders at sub-3d level.

**Comment by Marcin1960** [comment]
@betterworld "Add human games to the training database at this stage and continue reinforcement learning"

Are there many human games at LZ level?

**Comment by john45678** [comment]
Ray (https://github.com/zakki/Ray) can use LZ weights and has ladder code built in.

**Comment by fell111** [comment]
The initial network of 20x256 won't be parity with ELF network. So I don't think the winrate of 20 block network could be above 65%. If we have a 20 block network who has 60% winrate vs current best, that's enough for promotion.

**Comment by Ishinoshita** [comment]
I liked/s the experimental, explorational aspect of the project. So, even if strength is not an obsession, with a much bigger network (40×256), it could be assumed for a while that the network is not the bottleneck, only the RL pipeline. Make the network progress as expected might not be that easy, might need new ideas, more tweaks, experimentations, which would result in deeper understanding of the RL.

**Comment by betterworld** [comment]
@Marcin1960 
> Are there many human games at LZ level?

I don't think they need to be at LZ level. There are definitely some strategies and techniques that humans have discovered and the zero bots do not know, like fusekis, tesujis, or maybe even mirror go and how to prevent it. I am sure that LZ can find something to add to its repertoire that will make it stronger.

**Comment by Marcin1960** [comment]
Much more interesting than brute force approach, would be improvements of the algorithms. In this sense, the success of the second approach, ie getting into higher dan levels, could be more a curse than a blessing. So long as LZ was weak, the score hunters were on interested.

As Russians proverb goes "If we have strength, reason is not needed" (Sila yest, uma nie nada)

**Comment by Ishinoshita** [comment]
@Marcin: high level retains some fascination per se, not for the competition among tools but for the comparison with human go top level. As a go player, I would love to see answered the question "can a machine give Ke Jie 4 or 5 stones handicap". When you see how Elf games have lifted up LZ 15b, you can imagine that AZ 40b at 5200 Elo asymptot had still plenty of room for improvement.

But agree that this is not the scope of LZ project as I understand it and that the most interesting aspect of the project is algorithmic exploration.

**Comment by alreadydone** [comment]
Leela Master weights are trained with human games but show no improvement in strength and even regression (006772ff is one officially tested).

Golaxy reported giving three stones to second tier pros (in China) without komi and won all games.

**Comment by Marcin1960** [comment]
I tried 224x20 net - 62b5417b as LeelaZeroT on KGS with 4 dan handicap. The first thing it did was to walk into a ladder.

I will try dynamic komi version  v0.31 next

**Comment by featurecat** [comment]
Without dynamic komi, Leela zero can barely do 3 handicaps. You definitely need dynamic komi. Then it can even beat me with 5-6 stones. (5 loss, 6 I got bored but even game, 7 I won). I'm 3 dan

**Comment by alreadydone** [comment]
dynamic komi doesn't work with ELF.

**Comment by featurecat** [comment]
right... need to use an official Lz network

**Comment by Marcin1960** [comment]
@alreadydone "dynamic komi doesn't work with ELF."

What do you mean? I just did test run (Marcin vs LeelaZeroT) it and seemed to work.

Is it that it makes wrong calculations in middle of game?

**Comment by Marcin1960** [comment]
It seems to be running OK. Now against 3 dan RASKAN as LeelaZeroT no handicap.

I did run against as Marcin with 3 handicaps, a few minutes ago and it worked too.

Number of playouts is decent too.

**Comment by betterworld** [comment]
> Leela Master weights are trained with human games but show no improvement in strength

But afaik it did not use reinforcement learning after training from the human games

**Comment by Ishinoshita** [comment]
@betterworld: even the first version of AG (the very  first version that beat Fan Hui) applied RL after the SL phase from human games.

**Comment by Crazyeight101** [comment]
@betterworld The problem with human strategies is that they're already largely invalidated by the fact that they almost never win.  At best, you would be spending time teaching LZ strategies to avoid rather than searching for stronger ones.  Isn't one of the strengths of neural nets the ability to generalize and apply its learning to a variety of positions, which is why it can win in positions it hasn't encountered before?

We saw from AG Master's online series that humans who haven't trained with a zero bot can't even maintain an even position for 30 moves.

**Comment by qzq1** [comment]
I'm not sure whether this discussion indicates a clear path, but there seems to be some agreement that:
1. Rapid strength improvement is not in itself a central goal; 
2. Algorithmic exploration is important and interesting; and,
3. Making LZ a useful tool for go players is a significant consideration.

I find the techniques explored here fascinating.  As a knowledgeable, though weak (2k), go player myself, I still find the effectiveness of the approach astounding - almost magical - even as I understand more about it.  One of the things I find most intriguing is when there is an apparent change in playing style - e.g. changes in the 3-3 invasion josekis.  I wonder whether there is an AlphaZero "style" or if LZ will necessarily converge on AZ playing techniques.  Is LZ's playing approach at all contingent on the particular progression path, or does the RL technique necessarily move toward an optimum?

I don't know enough to judge whether any of these considerations argue for moving next to 20B or 30B or 40B.  And I don't know whether there is a substantial difference in coding effort (which I'm sadly ill-equipped to help with).  But my preference would be for more steps -- and so not to jump immediately to 40B -- if this creates more opportunity for algorithmic exploration and more opportunity to observe transitions in playing style.  Thanks again to everyone for your skill and effort on this project.

**Comment by alreadydone** [comment]
@Marcin1960 
> Can anything be done about?

No without retraining the network after fixing the value output afaik.
For the reason, see https://github.com/gcp/leela-zero/issues/1553#issuecomment-398649576 and https://github.com/gcp/leela-zero/issues/1599#issuecomment-402266422.

**Comment by Friday9i** [comment]
I'd go for 30x256, which has a lot of potential for improvement (contrary to 20x256 which is almost like ELF, and as we are not so far from ELF, the potential should be limited) and remains reasonably quick, especially with the move to v=1600.
Regarding the starting net to be used, it probably doesn't really matter as long as the learning rate is quite high (and we are not stuck with too many zeros in the network file, ie "dying ReLUs"). It may even be quite beneficial to start from a reasonably weaker net, as it may discover new concepts by itself, rather than being "stuck" to the current concepts and getting a hard time going past them?
Practically speaking: a 30x256 net was recently trained and tested, it didn't perform very well (about 100 Elo lower than current net LZ152 or LZ153 IIRC) but it could be a good starting point, as it should improve fast through selfplay and iterative learning, and it will probably become stronger than the current net after just 1 to a few weeks, with a huge margin of improvement thereafter!
Alternatively, we could try with @bjiyxo 20x256 v19 net: it would be almost as strong from the start, but it will probably be slower to improve and would stagnate quite quickly, as we are already close to it's maximum potential... Hence, that is not my preferred choice, I'd go for a larger net (30x256 probably).

And BTW, 2 technical questions: 
1. Do we had "dying ReLUs" issues, and if yes, how were they adressed? After a quick search here, it seems we didn't really encounter that problem as our net architecture is quite protected, isn't it? Thanks a lot.
2. Is it possible / difficult to test alternative network architectures? Eg introducing a global pooling layer as suggested in issue 1521? I guess it's quite hard, as the training pipeline has to be changed, isn't it?

**Comment by bjiyxo** [comment]
If we are going to explore something interesting and might cause radical changes, I suggest we should let the user choose which way they want to invest, and the default one should be the original way.

**Comment by jokkebk** [comment]
I don’t have a strong opinion (oe contribution to make my vote worth much, maybe a few hundred games done) but going 256x20 doesn’t rule out higher counts later, and at least the wisdom before has been small increments to maximize learning (more games generated). Of course if the new 256x40 is relly good, it might be possible as well, but if we get half the games, the progress might be slower than x20b first, the  40b

**Comment by featurecat** [comment]
@gondolieri But even the strongest Leela Zero network is still guessing the results of ladders. It must read them out, which means investing at least 40 playouts or so, which negatively impacts the search: since Leela Zero invests 3200 (or now 1600) visits per move and selects the move with the most playouts, it probably brings the overall strength of the network down. And Leela Zero *still* misreads ladders.

I would like a territory estimator...

**Comment by diadorak** [comment]
@Gondolieri , check @alreadydone 's work on dynamic komi.

**Comment by tterava** [comment]
> But even the strongest Leela Zero network is still guessing the results of ladders.

Calculating ladders is the reasonable thing to do. There's a limit how many things the network can learn based on its size, so learning to know the result of ladders based purely on policy network would be very taxing on it and would certainly make it weaker. It can very easily lead to game losing mistakes too, so you would have to be extremely accurate for it to be useful. Even human pros read the ladders out if they are not obvious.

Ladders really aren't a problem anymore as long as LZ is in a normal game state (no handicap, evaluation >10%). I still haven't seen an even game where LZ plays out a losing ladder to lose a game. So many people keep complaining about ladders, but I've only ever seen examples from handicap games for which LZ isn't designed for.

**Comment by gjm11** [comment]
@Gondolieri, I think what we want for an output is (in some form) a probability distribution over possible final scores. This should make it possible to play decently with arbitrary komi and in positions where merely winning is either nearly certain or nearly impossible. It shouldn't require a complete recapitulation of the learning process; rather, it would go something like this. 1. Augment the NN with a new "head" (or: change the existing "value head") that computes this probability distribution. 2. Train it on the self-play games with resignation disabled. 3. Train it, or perhaps now the whole network including the new head, on all self-play games, using the score predictions in the final position of each game as ground truth to train against for earlier positions in that game.

The "cheap" version of step 3 should be appreciably cheaper than redoing the whole learning process. The "expensive" version of step 3 would be just as bad as redoing the whole learning process, but would probably give better results (and I suspect might make the network play better too, even without adaptations to make use of the richer output data).

Then you'd modify MCTS to use not Pr(win) but Pr(final score difference >= x), with x chosen to be (1) the "correct" value, if that probability isn't too extreme, and otherwise (2) some value that pulls the probability into a more reasonable region.

What this wouldn't do is to make the policy network prefer riskier moves when behind and safer moves when ahead (or whatever the right thing for it to do actually is). For that, a current score or target score or something would need to be an _input_ to the network. As @diadorak suggests above, see @alreadydone's "dynamic komi" for one way to do that. At the moment it's a bit of a hack, but it should in principle be possible to make it less so -- e.g., by changing the self-play process to play games with different komi values.

All of this is super-speculative right now. It would be fun to try, though.

**Comment by qzq1** [comment]
Thanks @Crazyeight101.  Yes, I have seen the apparent convergence of playing technique, though I was holding out a vain hope that there might be an unexpected result.  I suppose the fact is that "style" is more relevant in the human realm, where other factors than pure analytical power are at play.  

To your point on strength improvement, yes, consistent improvement is a goal; not fastest improvement, per the comment of @gcp.

**Comment by TFiFiE** [comment]
Since it's already proving to be quite the challenge to create a 256x20 network that is stronger than the best 192x15 one (let alone under equal time control), it's hard to argue to even skip that size (BTW, what has actually gone into determining how to balance the number of blocks and channels in the past?). Regardless, I (also) think the next size should be the last to use the current architecture and that now is the time to decide what (backwards incompatible) ideas to incorporate into the next generation. My own vote for those right now would be:
1. (Dynamic) territory estimation (#98)
2. Global pooling (https://github.com/lightvector/GoNN#global-pooled-properties-dec-2017, https://github.com/gcp/leela-zero/issues/1129#issuecomment-377708501)
3. Whatever (possibly) comes out of #1109

I'm personally reluctant to see the "zero" approach abandoned in favor of ladder inputs.

**Comment by featurecat** [comment]
Ladders are just part of Go, similar to how we are using symmetry to speed up training. If "zero" means "no-human-ideas", ladders won't be non- zero because you literally can't play go without knowing how to calculate them. Forcing Leela Zero to calculate ladders out manually is like forcing your accountant to never use a calculator. It's going to slow them down for no reason, and they will make more mistakes.

**Comment by BITINT** [comment]
Yes to something like (Dynamic) territory estimation, yes to Global Pooling, and yes to other innovations such as but not limited to things like "FPU could be a trainable target" and "dynamic komi" and etc... At this stage its more important to shift from merely immitiating to starting to innovating. The world already has a fully open above-human level Go AI bot in the form of Leela Zero etc... its original mission has already well been accomplished. To do down the same route statically after its proven its point and after all the low hanging fruit is gone, post point of deminishing returns, in a post-ELF and possibly soon to be post-ELF v2.0 age, would be not the most interesting nor the most strategic pursuit of advancement further in the future.

**Comment by l1t1** [comment]
@Friday9i @bjiyxo 
maybe after https://github.com/gcp/leela-zero/issues/1573
the 40b networks  can be used for help trainning smaller weights, just like ELF did

**Comment by 22nsuk** [comment]
I agree to @TFiFiE 
In addition, can we consider testing the training methods of AZ before moving on to the next level?
It's probably a good time to test it out at 15 blocks before scaling up.

**Comment by Friday9i** [comment]
@l1t1 I doubt the 40b net is strong enough to train the current 15b net: to do that, we first should selftrain the 40b to a high level, then it will be possible to use these high level 40b games (including many moves unknown to the current 15b net) to improve the 15b net. Currently, to my knowledge, the 40b was only trained by games from the 15b net: I don't see how it could be useful to use it to train back the 15b net... After selftraining, possibly/probably, but before, no.
And also, some experiments were done to train smaller nets from current games, and they were not so successful in my understanding: the 64x5 net is much better than the best selftrained 64x5 but there was a bug in the program at that time, and tests on larger nets were not significantly better than best selftrained net of the same size. So, not clear it works: I expected it to work well, but that doesn't seem to be the case, or at least, nobody found a way to use it successfully yet.

**Comment by BITINT** [comment]
Since 15b has already went above and beyond its original call of duty, and regardless of what is the next size up, (20, 30, 40, etc) it will have to happen at some point or another, probably pretty soon... then why not lower the learning rate to as low as possible in order to squeeze as much strength/juice out of the remaining 15b as possible, for those that prefer a strong and relatively smaller sized network, this would be great to have right before moving up to the larger sizes.

**Comment by Tsun-hung** [comment]
@BITINT  Agree with you, but most people are out of patience, disapionted

**Comment by Friday9i** [comment]
Next steps are uneasy to choose...
Going on with a reduced learning rate to go to a top 15B net is a possibility, but it would take a lot of time and it won't go much further the point where we are already today. Hence, I don't think we should go this way (JMHO of course).
"Traditional alternatives": 
- Going for larger nets in order to get significantly stronger nets: as already stated, the 20x256 would probably be only sightly better, so I'd go directly to 30x256 (or even 40x256)
- Improving handicap play: that would be very nice, but it probably needs more time from developers, as it is probably not at all trivial to include and improve @alreadydone ideas in the mainstream LZ

"More experimental alternatives": trying new things! Eg:
- changing net structure (for example Global Pooling)
- including an MCTSnet in the LZ engine

Anyway, I'm confident the team of developers will do clever choices, depending on what they think is most important & feasible & promising : -). We'll know soon hopefully.

**Comment by Marcin1960** [comment]
@BITINT "At this stage its more important to shift from merely immitiating to starting to innovating. The world already has a fully open above-human level Go AI bot in the form of Leela Zero etc... its original mission has already well been accomplished."

Not yet I hope, there are still new ideas being developed and implemented. But do you have some innovation and ideas, you envision?

**Comment by Marcin1960** [comment]
@TFiFiE "Global pooling (https://github.com/lightvector/GoNN#global-pooled-properties-dec-2017, #1129 (comment))"

Fascinating, thanks.

**Comment by TFiFiE** [comment]
One thing that could be done right now with presumably little change is to "officially" generalize the color planes to komi ones and start training with different komi and corresponding inputs. I think a good range would be 5.5-8.5 (including the integers, for a total of 14 points to inter- and extrapolate from), as that should not lead to a lopsided advantage for one or the other color and significantly skewed test match results. This range is the largest one to contain only one odd number in order to (hopefully) help teach the network about parity with respect to area scoring.

**Comment by alreadydone** [comment]
I am also thinking about officially use the color planes as komi planes, though I am thinking of generating games with 2 handicap stones and 21.5 komi, etc. Due to current inaccurateness in winrate for komi other than +/-7.5, it may be preferable to use the nonslack("endgame") version to initalize the generation of such games. I hope +/-7.5 and +/-21.5 would provide enough data points to enforce a linear relationship between color plane inputs and komi, but maybe more komi values are needed.

For application in handicap games, the most important thing is monotonicity, and we could potentially be improved by data augmentation: if the side-to-move eventually loses the game it should lose the game under any higher komi. Sampling color plane inputs from a normal distribution centered around 0 or 1 for training the value head will likely lead to monotone and smooth winrate-komi relationship at large scale, while 5.5-8.5 komi training will detect jump discontinuities.

No matter how, once the net generalizes reasonably to other komi values, it would be possible to train another network with komi input closer to the outputs (to reduce computation for dynamic komi tuning etc.), by sampling positions from existing games and trying to predict outpus of the existing net. The new network can also incorporate a komi head with a winrate input which inverts the komi-to-winrate function (a sort of cumulative distribution function). That would solve the score estimation problem and also make dynamic komi tuning more straightforward.

**Comment by 5525345551** [comment]
I wonder if what @alreadydone proposes can be incorporated into leela zero without negatively affecting learning, so that people who want the strongest network possible can be happy, and people who want more innovation of features can be happy too.

I hope @gcp can consider it, mybe at the same time as the next net size increase. 

There are many ideas to try and probably only one experiment at a time is sensible. The ideas of @alreadydone  seems like a well thought out plan of development. Although I'm more of an enthusiastic supporter and can't code review etc...

Great work!

**Comment by Marcin1960** [comment]
"mybe at the same time as the next net size increase.
There are many ideas to try and probably only one experiment at a time is sensible."

In experiments, usually you change one factor at a time. For obvious reasons. It is more practical to test proposals of alreadydone now.

**Comment by l1t1** [comment]
leelaz has finished the matchs at https://github.com/gcp/leela-zero/issues/1554,
it seems leelaz is weaker than elf and aq in some condition, so both a stronger weight and a more efficient engine are needed in the future

**Comment by gcp** [comment]
>it seems leelaz is weaker than elf and aq in some condition

Some combination of less hardware, no tuning for the hardware used, and making the wrong moves. I don't think this is interesting at all, to be honest. Who has 10 x 1080Ti's?

**Comment by gcp** [comment]
>BTW, what has actually gone into determining how to balance the number of blocks and channels in the past?

I trained 2 networks for one of the past increases, raising filters in one and blocks in the other,  and then had a look at what preformed best. IIRC the result was that filters were surprisingly helpful. (I'd have bet on blocks myself)

**Comment by gcp** [comment]
>I hope @gcp can consider it

There is no flexible komi support in the main tree, so there's nothing to consider.

**Comment by gcp** [comment]
One note about trying varying komi: if you change the komi, you likely also change the correct policy moves in many situations. So if you are training to better handle various komi, you potentially lose raw playing strength compared to a net trained at fixed komi.

I'm not saying even more raw strength is necessarily more useful than better handling of handicap or alternate komi, just pointing out the potential trade-off.

**Comment by gcp** [comment]
>Hence, the approach "we wait for an equivalent net at time parity" may be a mistake...: there's little reason not to switch to a larger net, even if it is slightly worse at time-parity in the beginning.

If we fail to get a promotion at 192x15 and the training window fills up (I'll likely cut the ELF part in half or so as a last try), that issue sort of solves itself.

**Comment by alreadydone** [comment]
If @sbf2000 is willing to release training data, it would be easy to train a network that works under both 6.5 and 7.5 komi for 9x9, and compare strength with the single komi trained networks.

**Comment by gcp** [comment]
>recall at first it was just to test the bootstrapping and the method, not to grow to 15 block and looking to move to 20b or larger etc

I was going to stop at 128x6. I also wanted to stop when ELF came out, as there is now a publicly available super-human network. There was a good argument to continue, and try to get a slightly less stratified network closer in strength to ELF. We've gotten some progress there (still lacking strength), which improved handicap play and which has made the komi plane approach work, I assume.

Getting the strongest network isn't possible in the short/medium term. It's clear that more tightly integrated/more homogeneous groups with similar or more hardware can pump out stronger networks, and at best we're ensuring a baseline of freely available stuff. Longer term, this may not be true (open source projects can outlast commercial ones).

If you ask me what we're doing now it's getting margin so even weak hardware or no GPU gets high quality analysis.

In terms of codebase I'm happy with the current refactoring which will make the code easier for people who want to experiment. The only feature I still lack is setting arguments over GTP, and I'd love to have the OpenCL batching in the mainline tree, same for komi support.

Looking to the future, at some point I think it'll be necessary to have a production and a test side, split the clients say 80/20 and on the test side, run experiments where you restart from scratch and see how much (more) progress you can make in say 1-2-3 weeks compared to the baseline. Leela Zero Chess has been doing this, and I think it's a decent approach to try to find fundamental improvements.

Bonus points if as much as possible single points of failure (i.e. me) are eliminated.

> compare strength with the single komi trained networks.

I'm not sure if we have that, given that he used a trick to alternate komi as well.

**Comment by Friday9i** [comment]
Thanks for these precisions @gcp!

**Comment by john45678** [comment]
@gcp Firstly thank you for this great project. Are you still intending to upgrade to 20 Blocks?

**Comment by gcp** [comment]
>Are you still intending to upgrade to 20 Blocks?

Yes, 256x20 is still the plan for now, as soon as we have a suitable network. It's bootstrapping the latter that 's proving a bit difficult now. If there's no good 256x20 but we have a good 256x40 I'm certainly willing to just that that too.

**Comment by afalturki** [comment]
Why not 256x40 at 800 visits instead of 256x20 at 1600? We can produce games at almost the same rate but the progress should be faster since 256x40 is much farther from saturation.

**Comment by tapsika** [comment]
> Changes to network architecture such as Global Pooling are no-brainers: if you can train a stronger network than the current ones using the existing data, and, after adjusting for speed differences

> Why not 256x40 at 800 visits instead of 256x20 at 1600? We can produce games at almost the same rate

These two contradict each other. Training new kind of models to higher strength on existing data is only a reasonable expectation if that data is of high enough quality.

(Someone suggested earlier to have some games are played at high visits while others at low visits, which seems a reasonable compromise).

**Comment by afalturki** [comment]
> These two contradict each other. Training new kind of models to higher strength on existing data is only a reasonable expectation if that data is of high enough quality.

I don't understand how the two contradict each other in any way. I was not talking about existing data. After training the 40b network on existing data, the self-play progress for 40 blocks will probably be faster even with fewer visits because of the larger capacity of the network.

It would be great if the new 40b NN used Global Pooling instead of switching to GPP in the middle but that's not very related to my suggestion.

**Comment by gcp** [comment]
>Why not 256x40 at 800 visits instead of 256x20 at 1600?

Because we still use gating. If the gain is too small (because of using a small search), there might not be any promotions.

**Comment by gcp** [comment]
>Training new kind of models to higher strength on existing data is only a reasonable expectation if that data is of high enough quality.

If this is a concern, just try to make a stronger 128x10 with global pooling, using 192x15 data.

**Comment by TFiFiE** [comment]
> One note about trying varying komi: if you change the komi, you likely also change the correct policy moves in many situations. So if you are training to better handle various komi, you potentially lose raw playing strength compared to a net trained at fixed komi.
> 
> I'm not saying even more raw strength is necessarily more useful than better handling of handicap or alternate komi, just pointing out the potential trade-off.

Although "specialists" obviously always have a higher potential ceiling, training with both colors is effectively already like it's with different komi, so I question whether the loss of potential from a further generalization would be that major. It's very well possible that networks of current strength are already internally doing something close to assessing the risks and potential rewards of moves before making a choice based on a current score estimate, which is directly affected by the color/komi planes.

**Comment by Marcin1960** [comment]
"why not we make a new 18X384 network to upgrate? Since f is so important."

I would do it. Even 15X384 to isolate one factor. Double f and see.

**Comment by gcp** [comment]
>why not we make a new 18X384 network to upgrate? Since f is so important.

Sure, where can we download it to test?

**Comment by cjohnchen** [comment]
Thank you Marcin1960!

**Comment by l1t1** [comment]
gcp trained a 256x20 weight, the winrate to 157+ is 0.31 now

**Comment by Friday9i** [comment]
Yeah, not a big success in terms of winrate, but not too bad either: it's around 140 Elo below the last net. With time parity, it's probably around 180/200 Elo behind LZ157, but it has a significantly bigger potential.
Hence, I'd be tempted to test it's potential for a week in the main pipeline: if we get a good rythm of promotions from it (ie at least 3 or 4 promotions in a week), then we would be equivalent (on time parity) to LZ157 in less than 2 weeks. Indeed, a promotion is at least 35 Elo points, so we need 5 or 6 promotions to win back the 200 Elo points lost by switching to this net. And after these (less than) 2 weeks, because the potential is higher, we should continue to climb faster ; -).
And in the unfortunate case where we don't get frequent promotions, we can still revert to LZ157, having lost only 1 week: not a big problem IMHO.
What do you think of that?

**Comment by gcp** [comment]
I wish we had a table of all networks (including tests and failures) with their estimated Elo, it would make this discussion easier. @roy7

**Comment by Friday9i** [comment]
You get that info in the graph of tested nets, don't you? Not a table, but it gives the info nonetheless.

**Comment by gcp** [comment]
Yeah, it's just a missing feature I would find convenient.

Best so far is 48437672 (256x20 bjiyxo V19) at 11766 Elo.

**Comment by gcp** [comment]
I might try a new, larger bootstrap (say 256x30) with a larger window. Now that the Windows training support is fixed again, it's a bit easier to find a spare machine for this.

**Comment by jkiliani** [comment]
The early termination of failed tests due to SPRT is also somewhat inconvenient when comparing bootstrap nets. For example, bjiyxo V20 is certainly weaker than its reference 192x15 net, but the error bars of that measurement are far larger compared a test over 400 games.

**Comment by Friday9i** [comment]
Good point about v19, it is only 40 Elo below LZ157!
What do you think of trying it for about 1 week and see if it catches up with LZ157 on time parity through promotions?
Regarding GCP's net, it will be tested more (after ffd8 is finished), as it came back from fail and is now open again.

**Comment by gcp** [comment]
>What do you think of trying it for about 1 week and see if it catches up with LZ157 on time parity through promotions?

As said, at round ~200k game I'll fiddle with the ELF proportion in the training window. If that fails, there's no real other options but to force promote something, and force promoting the best larger network surely seems like the best bet.

**Comment by Friday9i** [comment]
Overlooked the approach 1) ELF 2) Force promote, makes sense of course, thanks for the precision.

Pity I'm not skilled enough to try training a net by myself, I'd be willing to try some alternative approaches (such as Global Pooling). I'm envisaging to find a computer expert and pay him to help me set-up everything and teach me how to do it in September...

**Comment by herazul** [comment]
@bjiyxo's 40b weights could be interesting too, we just have to wait 3 weeks to test it =)

**Comment by jkiliani** [comment]
> @bjiyxo's 40b weights could be interesting too, we just have to wait 3 weeks to test it =)

Do we actually? I though the point of holding back those 40 block nets was so they wouldn't be used against Leela Zero in the Tencent competition. Since LZ didn't make the final round anyway, is there still a reason not to share those weights?

**Comment by herazul** [comment]
he said he wanted to hold on to it until after this event : https://github.com/gcp/leela-zero/issues/1573

I don't know why or if he would reconsider it given that we might need it now, only him can answer that.

**Comment by Eddh** [comment]
> As said, at around ~200k games I'll fiddle with the ELF proportion in the training window. If that fails, there's no real other options but to force promote something, and force promoting the best larger network surely seems like the best bet.

I'm not clear on the details on the training right now. Are ELF games still being used in training ? If so it seems weird to me. In LeelaChessZero one one the thing that is being learned is that training on the same positions many time is bad. Decreasing sampling rate has produced good results for generalization every time it has been done.

**Comment by gcp** [comment]
>Are ELF games still being used in training ?

Yes. Half the window (250k + 250k). Reducing this is something I will try.

> Decreasing sampling rate has produced good results for generalization every time it has been done.

We tried removing them previously but the resulting networks were a little bit worse.

**Comment by Eddh** [comment]
How are the bootstraps being trained in regard to ELF games in particular ?

**Comment by gcp** [comment]
>How are the bootstraps being trained in regard to ELF games in particular ?

I tried without first and it was bad. I added them and it's...still bad, but maybe a bit better.

**Comment by Mardak** [comment]
> I wish we had a table of all networks (including tests and failures) with their estimated Elo, it would make this discussion easier

Not a table, but here's the data:

http://zero.sjeng.org/data/elograph.json

You can put this in your web console from a zero.sjeng.org page:
fetch("/data/elograph.json").then(r => r.json()).then(v => console.log(v.sort((a, b) => b.rating - a.rating)))

Here's the top by rating:
```
{"rating":11970,"net":8656658.1197,"sprt":"TEST","hash":"62b541","best":false}
{"rating":11886,"net":8245345.11886,"sprt":"TEST","hash":"62b541","best":false}
{"rating":11830,"net":8846456.1183,"sprt":"FAIL","hash":"fc5e0a"}
{"rating":11829,"net":8728634.11829,"sprt":"FAIL","hash":"f78ecb"}
{"rating":11821,"net":8814714.11821,"sprt":"FAIL","hash":"607f4a"}
{"rating":11820,"net":8846456.1182,"sprt":"FAIL","hash":"5be033"}
{"rating":11819,"net":8891928.11819,"sprt":"FAIL","hash":"ffd81b"}
{"rating":11816,"net":8891928.11816,"sprt":"FAIL","hash":"1fe10e"}
{"rating":11814,"net":8768394.11814,"sprt":"FAIL","hash":"502a07"}
{"rating":11808,"net":8891928.11808,"sprt":"FAIL","hash":"92c1f9"}
{"rating":11806,"net":8728634.11806,"sprt":"PASS","hash":"d351f0","best":true}
{"rating":11804,"net":8814714.11804,"sprt":"FAIL","hash":"1a06b5"}
```

**Comment by alphaladder** [comment]
The 20b test data of gcp and bjiyxo had suggested that a larger block(maybe 30?) is needed to exceed current 15b weight.

**Comment by Eddh** [comment]
if 20b can't even match 15b on equal nodes it suggest something wrong with the training to me. bootstraps were working before so I think it might come from 
1. temp=1 actually is harder to bootstrap on
2. It is harder to bootstrap using ELF games, or with no ELF games against a 15 block trained on ELF games.

**Comment by 22nsuk** [comment]
I'm supplementing the contents of the question and recapitulating.
I'm wondering about the test settings with the new 256x30 net.

1. You commented to test with a larger window, how many window does it mean?
2. What is the learing rate and batch size?

**Comment by gcp** [comment]
I'll try around 750k - 1M games. The second question makes no sense as the LR is reduced during training.

**Comment by l1t1** [comment]
force promoting the best larger network surely seems like the best bet.
--
why not force the best same size network( eg. fc5e0a50) ?

**Comment by 22nsuk** [comment]
Oh, I was talking about the initial learning rate.
And when bjioxyo trained 40 blocks from the beginning, I saw him start with 128 batch sizes.
But since 30 blocks is net2net, is it 256 batch size from the beginning?
I don't speak English well. Please understand me.
Thanks for the answer!

**Comment by alphaladder** [comment]
I know that the 40x256 weight which used in the 2018 Tencent World AI Weiqi Competition could beat 15b easily,which means that we could train a 40b network and run at v800or v1600.

**Comment by Mardak** [comment]
The concern about 800 visits is for future networks at the larger size. The visits is not for competing against a 15b. Self-play with more visits should generate training data that is X amount stronger, so that when a new network is trained, it learns to play at that X amount stronger level. Ideally that X amount stronger results in at least 55% win rate in 400 match games. However, if you reduce visits, we will likely generate training data that is less than X amount stronger, so it is less likely to pass the 55% threshold in test matches against the best network for promotion.

**Comment by alphaladder** [comment]
so what? 40b and v1600@22nsuk @mardak

**Comment by wonderingabout** [comment]
i think that regardless of what you want to do, a larger should grow stronger than a smaller one as time goes, whether it is 20 30 40 or whatever blocks
so i really think the first thing to do is to investigate and solve the cause of strength regressions on the 20b
force promoting seems quite ugly to me, and doesnt guarantee future problems wont appear

**Comment by fell111** [comment]
I doubt that whether we could get through the 55% gating if we decrease to 1,600 visits. Maybe we could try AZ approaching first? Then we can safely decrease visit to 1,600 or 800.

**Comment by anonymousAwesome** [comment]
> One note about trying varying komi: if you change the komi, you likely also change the correct policy moves in many situations. So if you are training to better handle various komi, you potentially lose raw playing strength compared to a net trained at fixed komi.
> 
> I'm not saying even more raw strength is necessarily more useful than better handling of handicap or alternate komi, just pointing out the potential trade-off.

I wouldn't be so sure about that.  From the abstract of the arxiv paper [Multi-Labelled Value Networks for Computer Go](https://arxiv.org/abs/1705.10701): 

>  First, the MSE of the ML [multi-labelled] value network is generally lower than the value network alone. Second, the program based on the ML value network wins by a rate of 67.6% against the program based on the value network alone. Third, the program with the proposed dynamic komi method significantly improves the playing strength over the baseline that does not use dynamic komi, especially for handicap games.

**Comment by Friday9i** [comment]
I don't see the problem with a larger but slightly weaker net...
It's sligthly weaker, true, but it has more potential for growth! So it should catch up with LZ157 quite qickly : -).
And regarding the v=1600, new nets should be tested with v=1600 against LZ157 also at v=1600 visits, so there should be no impact on gating, it's not a problem (still, I'd be happy if we test 50% gating, but that's another story).

**Comment by Marcin1960** [comment]
@Friday9i 
" (still, I'd be happy if we test 50% gating, but that's another story)."

I wonder, if the random fluctuations of wins were miraculously eliminated, wouldn't we see
a natural lowering of win numbers, WITHOUT stopping the improvements for a quite long time?

Perhaps net is not really saturated, only is not capable at high strength to make big jumps above 55%? Perhaps what is measured now is ability to make such big jumps?

If weaker net is promoted by accident, it should be eliminated quickly by the next net?

**Comment by l1t1** [comment]
let's bless the lz158 promoted soon. so that we have more time to discuss
![2018-07-25_165526](https://user-images.githubusercontent.com/23301704/43190264-323fb33a-902b-11e8-9d6c-5a58e7d3626a.png)

**Comment by Crazyeight101** [comment]
There's also the fact that we have, on average, about twice as many contributors and around 80% more games generated than we did a couple months ago.  So even if we go to a 30b or 40b that takes more time to generate self-play games and train, the rate at which we make progress shouldn't be much slower than what we were working with not too long ago.

@l1t1 lots of networks produce large winrates early in the series of test matches.  I've seen it happen dozens of times where it goes up to as much as 70% for the first 50-60 games, and then ends up somewhere between 48-52% by the end.  Don't hold your breath. =P

**Comment by Friday9i** [comment]
Yeah, one more... Still 60% after 85 games but around 49% after 230 games ;-(

**Comment by gcp** [comment]
>suggest to use elf weight as next best

It's not supported by the training code.

**Comment by l1t1** [comment]
can elf weight be convert to a leelaz traning codes supports size?

**Comment by higherdim** [comment]
@Crazyeight101 @Friday9i 
There is nothing mysterious about the winrate at relatively low number of games. 
For example, suppose the true expectation of winrate is 48%.   
For one game,  the chances of winrate higher than 55%, 60%  are 48%, 48% respectively,  of course! since it has 48% chance to win this single game (Wins / Losses 1:0). 
As the number of games increases, the winrate above 55% or 60% will decline, since it has the tendency to narrow down to 48%. 
For 50 games, the chances of winrate higher than 55%, 60%  are ~16%, ~6% respectively;
For 85 games, those are ~10.8%, ~1.75% ;
For400 games, those  are ~0.22%, ~0.0001%.

Nothing mysterious. Just statistics.

**Comment by diadorak** [comment]
The training window of 250K games is almost full. What would be the plan next? Manually promote fc5e0a50 to buy some time?

**Comment by wonderingabout** [comment]
@Friday9i 
you dont take into account many bias
i dont think that your distinction between short and long games indicates a different net behaviour. rather, the net is probably just following policie, and. if the phenomenon you're describing was to be true, you'd need a lot more data to assume it, and i suspect it would have very marginal impact on total winrate (arround 0.5-1%)
why is elf winrate higher at longer games ? not because lz is weaker in endgame, but probably just because elf would have resigned any game its losing much sooner, because of a different winrate evaluation from lz 156, on any board position
why is lz winrate higher at short games ? not because lz is stronger at short games, but probably just because elf resigns the games its losing early

also, every game is different : you can have 10 wins on a row then 10 losses on a row, because the games are played by a bot, it has no emotions and the previous result has no impact on next one, and even elf lost some games against the first 15b networks.
so, what matters is not a peculiar wining streak or a losing streak which are just random fluctuations, but the overall average on a big enough number of games, to have statistical trust (significance) in the average winrate. for example, elf may loose 5 games on a row against first 15b networks, but in 400 games it will have more than 90% winrate, the 5 losses on a row are not significant, but just random fluctuations

finally, i said it and i say it again, i think that the regression we observed on 20b training strength was not to be expected, as the 20b was supposed to be progressively stronger than the 15b and eventually replace it, and at least it shouldnt go weaker. i may be mistaken, but i think that as long no explanation (and fix) is found as of why 20b progressively lose strength, any forced promotion may promote the regression too, and possibly cause future problems

**Comment by wonderingabout** [comment]
P.S : 
a possible experiment can be to take 20bv15 network (which had 61% winrate against official lz), and forget about all the available training data that happened after 20bv15, and start using public ressource to train 20bv15 for 100-150k brand new games, and see if it gets stronger, or if it shows the same regression :
- if it gets stronger, keep training it for another new 100k games and see if it gets stronger. if it still gets stronger, keep training it for another new 100k and queue it against elf to see if it improved or not (you can call it 20b v15.1).
- however, if it gets weaker, go back to earlier versions of 20b an repeat the experience for 100k new games and see.

**Comment by Friday9i** [comment]
@higherdim: "event below 0.1% probability" is below 0.1% probability, not 4.4% probability... I'm speaking of very low probability I saw several times, such as a net leading 25-5 then going below 50% for example: that can happen, but that should be rare (around 0.01%!)! And it's not so rare, that's my first point (and the 2nd point is explaining why ;-).

@wonderingabout: I don't see your points on your first post above... I may not take into accounts many biais: possible, but what do you have in mind? "Why is elf winrate higher at longer games ?": who cares, it is just a fact (and the % cannot be due to luck, hard to compute but probability is very close to 0 for sure). You also say something around "what matters is the average winrate not the first games winrate": sorry, but we specifically speak about initial winrate vs final winrate, so it matters (you may not care of course, that's your right). All in all, if you do not agree be more specific please, rather than a vague "you don't take into account many biais"

**Comment by higherdim** [comment]
@wonderingabout 

>  i suspect it would have very marginal impact on total winrate

you are right since they are second order fluctuations that randomness fluctuations ( black-white randomness, long-short randomness...) multiply those specific variations (black-winrate difference, long-short winrate difference..).

**Comment by Friday9i** [comment]
I'm sorry but a 44% winrate on short games and 94% winrate on long games (data from ELF as posted above) has for sure a big impact on initial vs final winrate, because short games arrives on the server on average before long games.
No impact on the final %, of course, but an impact on the slope of the winrate %: it should be lower at the beginning then increasing at then end.

**Comment by wonderingabout** [comment]
like i said, you can win 15 games on a row then loose 30 games on a raw, its just random fluctuation

"For other nets, we observe the opposite, and for others, it seems undifferentiated."
in other words its just random fluctuation : 
- some start with high winrate then end low
- some start with low winrate then end high
- some start with high winrate and end high
- some start with low winrate and end low
- some start with low winrate and end middle (50%)
- some start with high winrate and end middle (50%)
i dont need to do the other ones do i ?

the first few games dont matter, even if its 90% then it ends at 40%, or the opposite, what matters is the final average on the total number of many games (400 in our setup) to be able to trust the significancy of our results. its just stats

**Comment by Friday9i** [comment]
30 games in a row with about 50% winrate has a probability of 1 in a billion. 15 games in a row a probability of 1 in 30000. So to get one after the other has a probability of about 1 in 3.10^13, which is totally ...0. Your random flucatuation example is just crap: sorry but you should be wonderingabout statistics before speaking about statistics...
And yeah, what matters **for promotion** (or not) is the end result, not the early result.
But if you didn't notice yet, I'll repeat: I'm speaking about **early match results versus end results**... That may not be interesting for you, no problem, but don't try to explain things you're not really understanding.

**Comment by Crazyeight101** [comment]
@higherdim I never suggested there was anything mysterious about it, and I'm very knowledgeable about statistics.  I wouldn't be able to work at my current job if I wasn't.

I was actually making more or less the same argument that you did, which is that it's not unusual to see very large win rates for smaller samples of games even when the new network is around the same strength as the current one.  I just didn't see a need to start rattling off std. errs and devs and CIs and how gating works to make my point...

**Comment by higherdim** [comment]
@Crazyeight101 Sure. I actually meant to reply to @l1t1 ...

**Comment by jkiliani** [comment]
@gcp: Were the training parameters (window size, ELF component) changed already for the current training run?

**Comment by alphaladder** [comment]
shall we try 30x256？

**Comment by john45678** [comment]
A bit off topic, Minigo has started their 19x19 v9 recently and already their best model (369 warrior) is scoring 50% against best LZ 128x10 net (39d46507).
see: http://www.cloudygo.com/v9-19x19/eval-graphs
and: http://cloudygo.com/cross-eval/eval-model/2523?sorted=True
discord: https://discordapp.com/channels/417022162348802048/433085673088614401

**Comment by MartinDevelopment** [comment]
@john45678 The sgf files state that minigo used around 6x more time than LZ.

**Comment by john45678** [comment]
@MartinDevelopment Just after noticing that now myself, @Seith didn't mention that in the discord post.

**Comment by diadorak** [comment]
And they have generated more than 8 million 20x256 games. I guess it means LZ's strategy of gradually increasing network size is more efficient?

**Comment by diadorak** [comment]
@MartinDevelopment  Minigo uses 20x256 network. So it should be much slower than LZ's 10x128 network.

**Comment by MartinDevelopment** [comment]
@diadorak That means the strenght of minigo's 20x256 network is nowhere near the strenght of the LZ 10x128 network, since it needs much more resources to be able to achieve a 50% result.

**Comment by diadorak** [comment]
Give them a few more weeks, they started this run last Friday...

**Comment by l1t1** [comment]
@gcp 

at ＃1113 said will force promote the best 20x256
but how to judge which one is the best?
elo isn't very exactly, the elf elo changes many times

**Comment by l1t1** [comment]
@gcp

at #1113 said will force promote the best 20x256, 
but how to judge which one is the best?
elo isn't very exactly, the elf elo has changed many times due to play against different weights.

**Comment by Mardak** [comment]
If we go by the 3-3 knight joseki.. V20-2 is best 😉 😉  https://github.com/gcp/leela-zero/issues/1442#issuecomment-389932483

```
256V18 S16 ->      48 (V: 52.91%) (N: 47.26%) PV: S16 R18 T17 Q18 S17 N18 O16 N17 R14 O14 Q12 C6 O15 N15 P15 N16 C12 Q13
256V19 S16 ->      33 (V: 53.39%) (N: 33.84%) PV: S16 R18 T17 Q18 S17 N18 O16 N17 R14 O14 Q12 C6 O15 N15 P15 N16
256V20 S16 ->      38 (V: 53.56%) (N: 38.07%) PV: S16 R18 T17 Q18 S17 N18 O16 N17 R14 O14 Q12 C6 C12 C3 C4 D3
256V20-2 S16 ->    63 (V: 52.85%) (N: 61.35%) PV: S16 R18 T17 Q18 S17 N18 O16 N17 R14 O14 Q12 C6 O15 N15 P15 N16 C12 Q13 C5 D6 F4
```

(Makes sense as ELF training data is what trains towards this "final joseki.")

**Comment by wonderingabout** [comment]
i'd be very curious if @ryouiki could perform a 20b v17 vs 20b v20-2 200 games match at 400 visits

**Comment by l1t1** [comment]
lz 158 is passed by sprt 0.4554

**Comment by 2ji3150** [comment]
If the current best weight is still weaker than the Elf weight ,even Elf weight with only 800 visit, Maybe it is better to use elf weight with 800 visit to boost up self-play generation ? Then we can catch up the stregh of the elf weight more faster.

**Comment by 2ji3150** [comment]
Elf OpenGo v1 is released. Seems slightly stronger.
https://github.com/pytorch/ELF/releases/tag/pretrained-go-19x19-v1
Maybe it is intresting to convert and match with it?

**Comment by l1t1** [comment]
elf v1
https://github.com/gcp/leela-zero/issues/1673

**Comment by Ishinoshita** [comment]
@gcp: simple curiosity: can you please tell us what 'black art' you applied to get the new 20x256 network out of the rut, if any? Learning rate tweaking? More?

**Comment by TFiFiE** [comment]
Knowing him, if he had done something, he would have probably mentioned it by now. I think it's just the natural result of the training window filling up with games created by the very neighborhood of networks which are trained on them.

**Comment by gcp** [comment]
The learning rate was lowered to 0.00005. I also increased the max steps to 512k. Not really what I was hoping, but given that no-one manged to produce a strong network using a higher rate, I guess we have to live with this? The last times I had a rate this low, it didn't produce further progress, but it does seem to have a positive effect now.

I'm off to the European Go Congress so I'll be less active for a few days. We'll have some idea whether 256x20 will approach 256x40 quickly or if  we should switch to the 256x40 by then.

**Comment by l1t1** [comment]
could you share the document of your lecture after the congress? @gcp

**Comment by alreadydone** [comment]
Just made a [release](https://github.com/alreadydone/lz/releases/tag/komi%2Bv0.4) (with very detailed readme!) out of my komi+next branch which is a stable, combined and updated version of komi-v0.3x and endgame-v0.3 and opened pull request #1772.
